daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PhaseOne.scala

```diff
@@ -504,168 +570,231 @@ private[lf] final class PhaseOne(
+        compileExp(env, cid) { cid =>
+          compileExp(env, arg) { arg =>
+            compileExp(env, trep) { trep =>
+              compileExp(env, guard) { guard =>
+                Return(t.GuardedChoiceDefRef(ifaceId, chId)(cid, arg, trep, guard))
+              }
+            }
+          }
+        }
+      case UpdateExerciseByKey(tmplId, chId, key, arg) =>
+        compileExp(env, key) { key =>
+          compileExp(env, arg) { arg =>
+            Return(t.ChoiceByKeyDefRef(tmplId, chId)(key, arg))
+          }
+        }
       case UpdateGetTime =>
-        SEGetTime
+        Return(SEGetTime)
       case UpdateLookupByKey(RetrieveByKey(templateId, key)) =>
-        t.LookupByKeyDefRef(templateId)(compile(env, key))
+        compileExp(env, key) { key =>
+          Return(t.LookupByKeyDefRef(templateId)(key))
+        }
       case UpdateFetchByKey(RetrieveByKey(templateId, key)) =>
-        t.FetchByKeyDefRef(templateId)(compile(env, key))
-
+        compileExp(env, key) { key =>
+          Return(t.FetchByKeyDefRef(templateId)(key))
+        }
       case UpdateTryCatch(_, body, binder, handler) =>
-        unaryFunction(env) { (tokenPos, env0) =>
-          s.SETryCatch(
-            app(compile(env0, body), env0.toSEVar(tokenPos)), {
-              val env1 = env0.pushExprVar(binder)
-              SBTryHandler(
-                compile(env1, handler),
-                env1.lookupExprVar(binder),
-                env1.toSEVar(tokenPos),
+        unaryFunction(env) { (tokenPos, env) =>
+          compileExp(env, body) { body =>
+            val env1 = env.pushExprVar(binder)
+            compileExp(env1, handler) { handler =>
+              Return(
+                SETryCatch(
+                  app(body, env.toSEVar(tokenPos)),
+                  SBTryHandler(
+                    handler,
+                    env1.lookupExprVar(binder),
+                    env1.toSEVar(tokenPos),
+                  ),
+                )
               )
-            },
-          )
+            }
+          }
         }
     }
 
   @tailrec
-  private[this] def compileAbss(env: Env, expr0: Expr, arity: Int = 0): s.SExpr =
-    expr0 match {
+  private[this] def compileAbss(env: Env, exp: Expr, arity: Int): Work = {
+    exp match {
       case EAbs((binder, typ @ _), body, ref @ _) =>
         compileAbss(env.pushExprVar(binder), body, arity + 1)
       case ETyAbs((binder, KNat), body) =>
         compileAbss(env.pushTypeVar(binder), body, arity + 1)
       case ETyAbs((binder, _), body) =>
         compileAbss(env.hideTypeVar(binder), body, arity)
       case _ if arity == 0 =>
-        compile(env, expr0)
+        compileExp(env, exp)(Return)
       case _ =>
-        withLabelS(t.AnonymousClosure, s.SEAbs(arity, compile(env, expr0)))
+        compileExp(env, exp) { exp =>
+          Return(withLabel(t.AnonymousClosure, SEAbs(arity, exp)))
+        }
     }
+  }
 
+  val compileAppsX = compileApps _
```

Comment:
I don’t understand this. Are you just trying to shut up the `@tailrec` check? I feel like dropping the `tailrec` annotation might be clearer for that.

Reply:
I guess @nickchapman-da  wants the of to handle `ETyApp` to be tail recursive but cannot make the one for `EApp` tail recursive. This looks fine for me as it is. 
A comment would be welcomed though.

Reply:
We only want to silence the `@tailrec` annotation in this one place. But we still want to ensure tail-recursion for the other call to `compileApps` in the `ETyApp` case.

Reply:
sorry, our comments overlapped

Reply:
I added a code comment to explain.